bgpd/bgp_vty.c

```diff
@@ -3600,6 +3600,57 @@ DEFUN (bgp_default_shutdown,
 	return CMD_SUCCESS;
 }
 
+DEFPY(bgp_shutdown_msg, bgp_shutdown_msg_cmd, "bgp shutdown message MSG...",
+      BGP_STR
+      "Enable administrative shutdown of the BGP instance\n"
+      "Add a shutdown message (RFC 8203)\n"
+      "Shutdown message\n")
+{
+	char *msgstr = NULL;
+
+	VTY_DECLVAR_CONTEXT(bgp, bgp);
+
+	if (argc > 3)
+		msgstr = argv_concat(argv, argc, 3);
+
+	bgp_shutdown_enable(bgp, msgstr);
+	XFREE(MTYPE_TMP, msgstr);
+
+	return CMD_SUCCESS;
+}
+
+DEFPY(bgp_shutdown, bgp_shutdown_cmd, "bgp shutdown",
+      BGP_STR "Enable administrative shutdown of the BGP instance\n")
+{
+	VTY_DECLVAR_CONTEXT(bgp, bgp);
+
+	bgp_shutdown_enable(bgp, NULL);
+
+	return CMD_SUCCESS;
+}
+
+DEFPY(no_bgp_shutdown, no_bgp_shutdown_cmd, "no bgp shutdown",
```

Comment:
can't we bundle like `no bgp shutdown [message MSG...]` here instead of using two commands?

Reply:
As mentioned in the summary of [my latest commit](https://github.com/FRRouting/frr/pull/6938/commits/1b6e7a8874bcd7c5a48d2c849e64b0dd67f49f7e), due to the fact that optional parameters and to-end-of-line string concatenation used in combination is currently not possible in vtysh command extraction, this seems to be the cleanest way  to do it.

Reply:
Thanks, sorry, didn't notice that in a commit message.

Reply:
I think we could save a few lines of code here by putting both commands under the same DEFPY like @ton31337 said.

Two nits here:
- The command description is normally "neutral" so I think the `Enable` or `Disable` in the description looks off. Normally one would find out what action the command would do by looking if there is a `no` in the beginning of the command or not.

   Example: `bgp shutdown` means that we are shutting down the instance and `no bgp shutdown` means we are lifting the shutdown condition.

   The descriptions would look like:
   `BGP information` for the `bgp` part, then `Administrative shutdown of BGP instance` for the `shutdown` part.

- Since the command logic is very simple in this case (e.g. no ellipsis concatenation or other things that need flow control) we could use `no bgp shutdown [message MSG...]` (the `[message MSG...]` is optional argument) to save a DEFUN declaration, code generation and yet another CLI command for vtysh token parser to match/find/list.

Reply:
Ops, disregard nit two I just saw the comments.

Reply:
Actually we can simplify it a bit further by using an `ALIAS`, check this out:
https://github.com/FRRouting/frr/blob/master/lib/filter_cli.c#L600

```c
DEFPY_YANG(
	no_access_list_remark, no_access_list_remark_cmd,
	"no access-list WORD$name remark",
	NO_STR
	ACCESS_LIST_STR
	ACCESS_LIST_ZEBRA_STR
	ACCESS_LIST_REMARK_STR)
{
	char xpath[XPATH_MAXLEN];

	snprintf(xpath, sizeof(xpath),
		 "/frr-filter:lib/access-list[type='ipv4'][name='%s']/remark",
		 name);
	nb_cli_enqueue_change(vty, xpath, NB_OP_DESTROY, NULL);

	return nb_cli_apply_changes(vty, NULL);
}

ALIAS(
	no_access_list_remark, no_access_list_remark_line_cmd,
	"no access-list WORD$name remark LINE...",
	NO_STR
	ACCESS_LIST_STR
	ACCESS_LIST_ZEBRA_STR
	ACCESS_LIST_REMARK_STR
	ACCESS_LIST_REMARK_LINE_STR)
```

Reply:
@rzalamena: I tried doing it both with aliases and with the optional parameters, but it produced either command extraction errors or even terminal crashes. The DEFPY_YANG does not seem to be mentioned in the dev guide, did i miss some new thing there?

Reply:
`DEFPY_YANG` is a new DEFPY for YANG commands, it only differs from regular DEFPY to clue northbound that it can be batched when not using transactional cli.

I tried the `ALIAS` approach and it works for me:

```diff
diff --git a/bgpd/bgp_vty.c b/bgpd/bgp_vty.c
index db0a5820f..86bbd4c87 100644
--- a/bgpd/bgp_vty.c
+++ b/bgpd/bgp_vty.c
@@ -3629,8 +3629,11 @@ DEFPY(bgp_shutdown, bgp_shutdown_cmd, "bgp shutdown",
        return CMD_SUCCESS;
 }
 
-DEFPY(no_bgp_shutdown, no_bgp_shutdown_cmd, "no bgp shutdown",
-      NO_STR BGP_STR "Remove administrative shutdown of the BGP instance\n")
+DEFPY(no_bgp_shutdown, no_bgp_shutdown_cmd,
+      "no bgp shutdown",
+      NO_STR
+      BGP_STR
+      "Administrative shutdown of the BGP instance\n")
 {
        VTY_DECLVAR_CONTEXT(bgp, bgp);
 
@@ -3639,17 +3642,13 @@ DEFPY(no_bgp_shutdown, no_bgp_shutdown_cmd, "no bgp shutdown",
        return CMD_SUCCESS;
 }
 
-DEFPY(no_bgp_shutdown_msg, no_bgp_shutdown_msg_cmd,
-      "no bgp shutdown message MSG...", NO_STR BGP_STR
-      "Remove administrative shutdown of the BGP instance\n"
-      "Add a shutdown message (RFC 8203)\n" "Shutdown message\n")
-{
-       VTY_DECLVAR_CONTEXT(bgp, bgp);
-
-       bgp_shutdown_disable(bgp);
-
-       return CMD_SUCCESS;
-}
+ALIAS(no_bgp_shutdown, no_bgp_shutdown_msg_cmd,
+      "no bgp shutdown message MSG...",
+      NO_STR
+      BGP_STR
+      "Administrative shutdown of the BGP instance\n"
+      "Add a shutdown message (RFC 8203)\n"
+      "Shutdown message\n")
 
 DEFUN (neighbor_remote_as,
        neighbor_remote_as_cmd,
```

Reply:
No terminal crash?

Reply:
No terminal crash:

```
frr-debian(config-router)# bgp shutdown message foo
frr-debian(config-router)# no bgp shutdown message foo
frr-debian(config-router)# bgp shutdown
frr-debian(config-router)# no bgp shutdown
```

BGP logs:
```
2020/08/24 09:48:06 BGP: Enabled administrative shutdown on BGP instance AS 100
2020/08/24 09:48:09 BGP: Disabled administrative shutdown on BGP instance AS 100
2020/08/24 09:48:20 BGP: Enabled administrative shutdown on BGP instance AS 100
2020/08/24 09:48:24 BGP: Disabled administrative shutdown on BGP instance AS 100
```

Try the diff and see for yourself. I think you might have gotten the wrong first `ALIAS` argument (it is supposed to be the command function you want to replicate).

Reply:
@rzalamena: Does [this commit](https://github.com/FRRouting/frr/pull/6938/commits/9ddf4b8180879b982f6764b06884108689e014ce) resolve your concerns regarding the command description and use of an ALIAS definition?

Reply:
@davischw yes it does, thank you!